src/github/mod.rs

```diff
@@ -0,0 +1,220 @@
+#![allow(unused_must_use)]
+use crate::{
+    embeddings::{Embeddings, EmbeddingsModel},
+    prelude::*,
+};
+use rayon::prelude::{IntoParallelIterator, ParallelIterator};
+use serde::{Deserialize, Serialize};
+use std::io::Read;
+
+#[derive(Debug, Default, Serialize)]
+pub struct File {
+    pub path: String,
+    pub content: String,
+    pub length: usize,
+}
+
+impl ToString for File {
+    fn to_string(&self) -> String {
+        format!(
+            "File path: {}\nFile length: {} bytes\nFile content: {}",
+            &self.path, &self.length, &self.content
+        )
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct FileEmbeddings {
+    pub path: String,
+    pub embeddings: Embeddings,
+}
+
+#[derive(Debug)]
+pub struct RepositoryEmbeddings {
+    pub repo_id: String,
+    pub file_embeddings: Vec<FileEmbeddings>,
+}
+
+#[derive(Serialize, Debug)]
+pub struct RepositoryFilePaths {
+    pub repo_id: String,
+    pub file_paths: Vec<String>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Repository {
+    pub owner: String,
+    pub name: String,
+    pub branch: String,
+}
+
+impl ToString for Repository {
+    fn to_string(&self) -> String {
+        format!("{}-{}-{}", &self.owner, &self.name, &self.branch)
+    }
+}
+
+pub async fn embed_repo<M: EmbeddingsModel + Send + Sync>(
+    repository: &Repository,
+    files: Vec<File>,
+    model: &M,
+) -> Result<RepositoryEmbeddings> {
+    let file_embeddings: Vec<FileEmbeddings> = files
+        .into_par_iter()
+        .filter_map(|file| {
+            let embed_content = file.to_string();
+            let embeddings = model.embed(&embed_content).unwrap();
+            Some(FileEmbeddings {
+                path: file.path,
+                embeddings,
+            })
+        })
+        .collect();
+
+    Ok(RepositoryEmbeddings {
+        repo_id: repository.to_string(),
+        file_embeddings,
+    })
+}
+
+pub async fn fetch_repo_files(repository: &Repository) -> Result<Vec<File>> {
+    let Repository {
+        owner,
+        name,
+        branch,
+    } = repository;
+
+    let url = format!("https://github.com/{owner}/{name}/archive/{branch}.zip");
+    let response = reqwest::get(url).await?.bytes().await?;
+
+    let reader = std::io::Cursor::new(response);
+    let mut archive = zip::ZipArchive::new(reader)?;
+
+    let files: Vec<File> = (0..archive.len())
+        .filter_map(|file| {
+            let mut file = archive.by_index(file).unwrap();
+            let file_path = file.name().split_once('/').unwrap().1.to_string();
```

Comment:
Looks like this will index every file, every single for repos that have already been processed.

I noticed this when hitting my local deployment with the same embedding request:
```json
{
	"owner": "open-sauced",
	"name": "insights",
	"branch": "beta"
}
```
Each request took about 6-7 seconds.

In the future, it'd be really interesting to only re-index files that have changed since the last time the repo was indexed. We could store the last time the repo was indexed and then iterate files in the git tree that have changed since that last embedding on that branch took place.

Reply:
Right. If we introduce a relational DB to the project we can implement more of such improvements, by maintaining info about the commit for which the repo was indexed etc and compare the changes.